qiime2/sdk/usage.py

```diff
@@ -159,80 +431,160 @@ def build_opts(self, action_signature, scope):
-    def source(self):
-        return self._source
+        if action_id == '':
+            raise ValueError('Must specify a value for action_id.')
 
-    def assert_has_line_matching(self, label, path, expression):
-        return self._assert_has_line_matching_(self.ref, label, path,
-                                               expression)
+        self.plugin_id = plugin_id
+        self.action_id = action_id
+        self._plugin_manager = sdk.PluginManager()
 
+    def __repr__(self):
+        return 'UsageAction(plugin_id=%r, action_id=%r)' % \
+               (self.plugin_id, self.action_id)
+
+    def get_action(
+        self,
+    ) -> typing.Tuple[
+        typing.Union[sdk.Method, sdk.Pipeline],
+        typing.Union[MethodSignature, PipelineSignature],
+    ]:
+        """
+        Get this example's action and signature.
+
+        Returns
+        -------
+        action_f : QIIME 2 Method or Pipeline
+            The plugin action
+        action_f.signature: QIIME 2 Method or Pipeline signature
+            The method signature for the plugin action
+        """
 
-class Scope:
-    def __init__(self):
-        self._records = dict()
+        plugin = self._plugin_manager.get_plugin(id=self.plugin_id)
+        # TODO: should this validation be pushed up into
+        # plugin.py or action.py?
+        try:
+            action_f = plugin.actions[self.action_id]
+        except KeyError:
+            raise KeyError('No action currently registered with '
+                           'id: "%s".' % (self.action_id,))
+        return action_f, action_f.signature
 
-    def __repr__(self):
-        return '%r' % (self._records, )
+    def validate(self, inputs: UsageInputs, outputs: UsageOutputNames) -> None:
+        """
+        Validate inputs and outputs for this example.
 
-    @property
-    def records(self):
-        return types.MappingProxyType(self._records)
+        Parameters
+        ----------
+        inputs : UsageInputs
+        outputs : UsageOutputNames
 
-    def push_record(self, ref, value, source, assert_has_line_matching=None):
-        record = ScopeRecord(ref=ref, value=value, source=source,
-                             assert_has_line_matching=assert_has_line_matching)
-        self._records[ref] = record
-        return record
+        """
+        if not isinstance(inputs, UsageInputs):
+            raise TypeError('Must provide an instance of UsageInputs.')
+        if not isinstance(outputs, UsageOutputNames):
+            raise TypeError('Must provide an instance of UsageOutputNames.')
 
-    def get_record(self, ref):
-        try:
-            return self.records[ref]
-        except KeyError:
-            raise KeyError('No record with ref id: "%s" in scope.' % (ref, ))
+        _, sig = self.get_action()
+
+        inputs.validate(sig)
+        outputs.validate(sig)
 
 
 class Usage(metaclass=abc.ABCMeta):
+    """`Usage` is the base class for Usage driver implementations.
+    """
```

Comment:
```suggestion
    """
    `Usage` is the base class for Usage driver implementations.
    """
```

Reply:
Is there a practical reason to prefer this suggestion over my original commit?  I don't see the preference recommended anywhere, including PEP 257, https://numpydoc.readthedocs.io/en/latest/format.html.  I also see both variations in the framework source code as well as NumPy source code.  I don't really care, I'm just curious!  😄 .  I think I actually prefer your suggestion.  I was just mimicking the examples in PEP 257 (which does mention either variation is fine).  Just curious if there is a deeper practical reason I'm unaware of.

Reply:
> Is there a practical reason to prefer this suggestion over my original commit?

Nope, you can ignore it - it was part of a broad set of linting rules I applied locally. I have no preference, but personally I like the `"""\nfoo\n"""` style, I think the `"""` delimits the comments well.

> Just curious if there is a deeper practical reason I'm unaware of.

I think the only other thing to keep in mind with these kinds of decisions is project/repository consistency. I quick spin around this repo shows examples of both docstring styles, so no real pattern here, but something to keep in mind, in general.

Reply:
>Nope, you can ignore it - it was part of a broad set of linting rules I applied locally.

Out of curiosity, do you apply those programmatically somehow (YAPF, IDE functionality, etc.)?